### PR TITLE
Do not use fixed index in the `TreeNode.PrevNode` if the `TreeView` is sorted 

### LINKED ISSUE #8768: TreeNodeCollection.AddRange works incorrectly for non-empty collection of the sorted TreeView
### .NET version

6.0

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No

### Issue description

When the tree nodes are appended to the non-empty `TreeNodeCollection` owned by the sorted `TreeView` with already created underlying window, the resulting order of items in the actual window does not match the one in the collection (which is correct).

The cause is how method `TreeNode.PrevNode` works: when called from within `TreeNodeCollection.AddRange` of non-empty collection, it always uses fixed index as the position of the previous node, assuming the normal backward insertion of the nodes in the unsorted tree. Yet in case of the sorted `TreeView`, insertions are made at various positions according to a sort order.

There obviously should be a check for the sorted state of the tree in `TreeNode.PrevNode` to avoid this.

### Steps to reproduce

```
namespace TreeViewTest
{
  internal static class Program
  {
    [STAThread]
    static void Main()
    {
      ApplicationConfiguration.Initialize();

      var treeView = new TreeView();
      treeView.Nodes.Add(new TreeNode("2"));
      treeView.Sort();

      var form = new Form();
      form.Controls.Add(treeView);
      form.Load += (sender, e) => treeView.Nodes.AddRange(new TreeNode[] {
        new TreeNode("3"),
        new TreeNode("5"),
        new TreeNode("1"),
        new TreeNode("4")
      });
      Application.Run(form);
    }
  }
}
```

This results in:
Clipboard01

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #8768

This is a re-roll of the PR in #8774, but with the switch added.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/10493)

### CODE REVIEW

**Comment by elachlan** [comment]
CC: @snnz

File: src/System.Windows.Forms.Primitives/src/System/LocalAppContextSwitches/LocalAppContextSwitches.cs [hunk 1]
```diff
@@ -165,5 +167,11 @@ public static bool DataGridViewUIAStartRowCountAtZero
         get => GetCachedSwitchValue(DataGridViewUIAStartRowCountAtZeroSwitchName, ref s_dataGridViewUIAStartRowCountAtZero);
     }
 
+    public static bool TreeNodeDoNotUseFixedIndexWithSortedTreeView
```
Comment (lonitra) [inline src/System.Windows.Forms.Primitives/src/System/LocalAppContextSwitches/LocalAppContextSwitches.cs:170]: Please add comments for this switch

**Comment by elachlan** [comment]
@lonitra I added the xml comment and renamed the switch to be more specific to its behavior.

**Comment by snnz** [comment]
But this re-roll, it is totally incorrect. The purpose was to make sure that fixed index is NOT used if the tree is sorted, not vice versa. See the original PR: there should be `!TreeView.Sorted` in the condition.

**Comment by lonitra** [comment]
> @lonitra I added the xml comment and renamed the switch to be more specific to its behavior.

Thank you! The change looks good. Could you add tests to capture the behavior of the switch when it is on?

**Comment by elachlan** [comment]
Sorry! I'll try and get this sorted soon.

**Comment by elachlan** [comment]
@lonitra I am unsure how we can unit test this. As the behavior of affects the underlying win32 tree view control.

**Comment by elachlan** [comment]
I did some manual testing and was getting different results than what was specified in the #8768.

I then looked into the differences, we changed the behavior of the insert in #9078 by @gpetrou 

Did we want to change the behavior back, or keep it as is? It used to add in "almost" reverse order, but now it adds it in the order provided (almost).

**Before:**
223826227-6cf6d1aa-3605-4e39-98a8-65116d5db21a

**After:**
image

Method of adding:
```csharp
  treeView1.Nodes.AddRange([
  new TreeNode("3"),
  new TreeNode("5"),
  new TreeNode("1"),
  new TreeNode("4")
]);
```

**Comment by elachlan** [comment]
Test is showing some weird behavior and I am not sure why. It doesn't line up with what I observed when manually testing.

Expected: `2,5,3,1,4`
Actual:  `2,3,4,5,1`

**Comment by lonitra** [comment]
> Did we want to change the behavior back, or keep it as is? It used to add in "almost" reverse order, but now it adds it in the order provided (almost).

Had you investigated what exactly caused the behavior change? From glancing at the PR it doesn't seem like the change was intentional.

> Test is showing some weird behavior and I am not sure why. It doesn't line up with what I observed when manually testing.

Hm.. That is concerning. It seems like the state between our control and the underlying one is mismatched and the results you are getting are actually being sorted when it comes to the underlying control.. This looks a bit suspicious to me
https://github.com/dotnet/winforms/blob/226a5b5d3782f40d8c84cdacfa3c49b3f7cc042e/src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNodeCollection.cs#L315-L318
this is where it looks like we eventually get to from `AddRange`. Could you maybe debug through this and see if somewhere along the line we are adding the tree node to a unexpected index for the underlying control? -- I cannot look into this too deepy right now, apologies!

**Comment by elachlan** [comment]
Sorry I haven't sorted this yet. Hopefully soon.

File: src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNode.cs [hunk 2]
```diff
@@ -812,17 +813,20 @@ public TreeNodeCollection Nodes
...
+            int currentIndex = _index;
+            int fixedIndex = _parent.Nodes.FixedIndex;
 
-            if (fixedInd > 0)
+            if (fixedIndex > 0)
             {
-                currentInd = fixedInd;
+                if (!LocalAppContextSwitches.TreeNodePrevNodeDoNotUseFixedIndexWithSortedTreeView || TreeView is null || !TreeView.Sorted)
+                {
+                    currentIndex = fixedIndex;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNode.cs:823]: The idea of this fix is that we want to skip the fixed index perf optimization when we are sorting the treeview.

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNode.cs:823]: Yes, based on this comment from the issue reporter.
https://github.com/dotnet/winforms/pull/10493#issuecomment-1873621647

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNode.cs:823]: Then we don't have to set FIxedIndex to the currently displayed number of child nodes, I prefer fix in https://github.com/dotnet/winforms/pull/11300

Comment (elachlan) [reply src/System.Windows.Forms/src/System/Windows/Forms/Controls/TreeView/TreeNode.cs:823]: No problem. Ill close this then. The tests still seem weird. So worth investigating.

**Pull request closed without merging**